docs/proposals/20230712-resource-limit-aware-scheduling.md

```diff
@@ -0,0 +1,146 @@
+
+// keyed by resourceName, value is ratio percentage
+type LimitToAllocatableRatio map[corev1.ResourceName]int64
+
+func GetNodeLimitToAllocatableRatio(annotations map[string]string, defaultLimitToAllocatableRatio LimitToAllocatableRatio) (LimitToAllocatableRatio, error) {
+	data, ok := annotations[AnnotationLimitToAllocatableRatio]
+	if !ok {
+		return defaultLimitToAllocatableRatio, nil
+	}
+	limitToAllocatableRatio := LimitToAllocatableRatio{}
+	err := json.Unmarshal([]byte(data), limitToAllocatableRatio)
+	if err != nil {
+		return nil, err
+	}
+	return limitToAllocatableRatio, nil
+}
+```
+
+The plugin configuration is as follows:
+
+```go
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// LimitAwareArgs defines the parameters for LimitAware plugin.
+type LimitAwareArgs struct {
+	metav1.TypeMeta
+
+	// Resources a list of pairs <resource, weight> to be considered while scoring
+	// allowed weights start from 1.
+	Resources []schedconfig.ResourceSpec
+
+	DefaultLimitToAllocatableRatio extension.LimitToAllocatableRatio
+}
+```
+
+### requested limit of pod
+
+Similar to calculating the resource request of a Pod, we use the following function to calculate the requested limit of a Pod. When the container does not specify Limits, use Requests as the limit of the container
+
+```
+func getPodResourceLimit(pod *corev1.Pod) *framework.Resource {
+	result := &framework.Resource{}
+	for _, container := range pod.Spec.Containers {
+		result.Add(quotav1.Max(container.Resources.Limits, container.Resources.Requests))
+	}
+	// take max_resource(sum_pod, any_init_container)
+	for _, container := range pod.Spec.InitContainers {
+		result.SetMaxResource(quotav1.Max(container.Resources.Limits, container.Resources.Requests))
+	}
+	// If Overhead is being utilized, add to the total limits for the pod
+	if pod.Spec.Overhead != nil {
+		result.Add(pod.Spec.Overhead)
+	}
+	return result
+}
+```
+
+### filter out nodes with high risk of resource over-subscription
+
+From before, we can already determine the Allocatable Limit and Allocated Limit of Node. When scheduling a new Pod, all nodes which satisfy requested limit + allocated limit > allocatable limit will be filtered out
+
+### score nodes to favor less risk of resource over-subscription
+
+Since the main purpose of the LimitAware Score plugin is to reduce the risk of resource over-subscription through Spreading or Balancing, this proposal uses the Least Allocated Limit strategy to score nodes.  We calculate each node’s raw score by this formula: `Score = (allocatable limit - allocated limit) * MaxNodeScore / allocatable limit` . For multiple resources, the raw score is the weighted sum of all allocatable resources. The resources and their weights are configurable as the plugin arguments. Like the other `NodeResources` scoring plugins, the plugin supports standard resource types: CPU, Memory, Ephemeral-Storage plus any Scalar-Resources.
```

Comment:
We need to redesign a new scoring algorithm. The reason is that the sum of pod limits is indistinguishable by the leastAllocated algorithm in the scenario where the maximum limit of the node is not set by default.